File: user/lib/audio.lua
-- Kosmos. Copyright (c) 2026 Diego Cibils. MIT; see LICENSE.
-- Playing a sound, from a program that does not own the sound device.
--
-- Almost nothing owns it: `SPAWN_AUDIO` goes to one process, the audio
-- server, and that is what makes per-application volume a thing that can
-- exist. If every program wrote to the device the last one would win, and
-- there would be nothing for a volume control to be a volume *of*.
--
-- So a program opens a *stream*, says what it is called - the name is what
-- appears in the Mixer - and sends periods. What comes back from `play` is
-- whether the server took it: false means the stream is already as far
-- ahead as it is allowed to be, and the caller should wait rather than
-- buffer, because the queue is the latency.

local audio = {}

local OP_OPEN, OP_CLOSE, OP_SET, OP_STREAMS = 1, 2, 3, 4

--
-- The format, asked rather than assumed.
--
-- `sys.info()` reports what the *board* does, and this passes it on so a
-- caller can size its buffers without a constant of its own. Zero when the
-- machine has no sound device, which is a supported way to run.
--
function audio.format()
  local i = sys.info() or {}

  return {
    rate = i.audio_rate or 0,
    channels = i.audio_channels or 0,
    period = i.audio_period or 0,
    periods = i.audio_periods or 0,
    frames = (i.audio_period or 0) // (2 * math.max(1, i.audio_channels or 1)),
  }
end

local stream = {}
stream.__index = stream

--
-- Returns whether the server took it, and *why not* when it did not.
--
-- The second value matters more than it looks. `play` returning false has
-- two meanings - the stream is full, which is normal and means wait, and
-- the message did not arrive, which is a bug - and a caller that cannot
-- tell them apart yields for ever on the second one. That is exactly what
-- happened: two tones sat in the Mixer at "idle" while `beep` spun,
-- because a failed send and a full queue looked identical.
--
--
-- Hand over one period.
--
-- **Nothing is sent.** The samples go into the ring this stream opened
-- with, which the audio server is already reading from - `CLAUDE.md`'s
-- rule, and `audioring.h` has the reasoning. What used to happen here was
-- an IPC call with 1024 bytes of payload, 172 times a second, minting a
-- Lua string on each side of every one.
--
-- Returns false with "full" when the ring has no room, which means this
-- client is ahead of the device and should do something else for a moment.
-- That is a fact rather than a failure, and it is the only thing keeping
-- the pipe short.
--
function stream:play(pcm)
  if sys.ring_put(self.ring, pcm) then return true end

  return false, "full"
end

--
-- The same, but wait for room.
--
-- For a program whose whole job is playing - `play` at a prompt - as
-- against one with a window to keep answering, which wants `play` and its
-- "full" so it can go and do that instead. Two shapes of caller, not two
-- ways of doing one thing.
--
function stream:write(pcm)
  --
  -- Bounded, because the alternative is a program that never returns.
  --
  -- This waited for ever. If the server stops draining the ring - because
  -- it died, and a server spawned with one capability dies *silently* -
  -- then every client blocks in here with nothing on screen and nothing on
  -- the serial line. That is indistinguishable from a hung machine, and it
  -- cost an evening of suspecting the test harness.
  --
  -- Two seconds is far longer than any legitimate wait: the ring plus the
  -- device is well under a hundred milliseconds of sound, so anything past
  -- that is not back pressure, it is a fault.
  --
  for _ = 1, 500 do
    if sys.ring_put(self.ring, pcm) then return true end

    -- One tick. The ring holds several periods and the device a few more,
    -- so there are tens of milliseconds still to play while this sleeps.
    sys.sleep(1)
  end

  return false, "the audio server stopped taking periods"
end

-- How many periods are waiting to be mixed, and how much room is left.
function stream:queued()
  return sys.ring_ready(self.ring) or 0
end

function stream:space()
  return sys.ring_space(self.ring) or 0
end

--
-- Where this stream has actually got to, in frames out of the speaker.
--
-- Not the same as what has been handed over, and the gap between the two is
-- the point: the ring holds periods the server has not taken yet and the
-- device holds periods it has taken and not played. A position built from
-- what this process knows would be ahead by both of them - about seventy
-- milliseconds - which is enough to put a progress bar visibly wrong and
-- far too much to sync anything to.
--
-- The server publishes it into this stream's own ring, so reading it is a
-- load rather than a round trip, and `music` can ask on every pass without
-- thinking about it.
--
function stream:position()
  local played = sys.ring_position(self.ring)

  return played or 0
end

--
-- How many frames between a sample written now and that sample being heard.
--
-- This is the number to divide by `self.rate` for a latency in seconds, and
-- it is the honest one: it counts the ring and the device together, because
-- it is a subtraction of two frame counts rather than a sum of two things
-- somebody has to remember to add.
--
-- Zero before anything has been written, and zero rather than negative if
-- the two halves are read across an update.
--
function stream:delay()
  local played, written = sys.ring_position(self.ring)
  local ahead = (written or 0) * self.frames - (played or 0)

  if ahead < 0 then return 0 end

  return ahead
end

--
-- The worst latency this stream can ever have, in frames.
--
-- The ring plus the device, because those are the only two places a frame
-- can be waiting. `delay` is what the latency *is* at this instant; this is
-- what it cannot exceed, and the gap between them is slack the client has
-- chosen not to use.
--
-- Worth having as a number rather than a comment: it is what a caller
-- checks its own `delay` against, and a delay past the bound means the
-- position and the indices disagree rather than that the machine is busy.
--
function stream:bound()
  return (self.ring_periods + self.periods) * self.frames
end

--
-- Declared here and defined in the protocol block below, because `open` and
-- `close` need it and the layout it depends on reads better next to the
-- rest of the wire format than scattered up here.
--
local request

function stream:close()
  request(OP_CLOSE, { stream = self.id })

  --
  -- The ring is this process's memory and the server only borrowed a view.
  -- It unmaps and drops its own name for it on `close`; this drops ours.
  --
  if self.cap then sys.release(self.cap) end
end

--
-- A stream, named for the Mixer.
--
-- The name is the program's own and nothing checks it: two copies of the
-- same program get two streams with the same name, which is right - they
-- are two things making a noise and each should have its own fader.
--
-- **The client creates the ring and hands over a capability to it**, the
-- same way a window hands the compositor its surface. The server can reach
-- that ring and nothing else in this process, which is what makes shared
-- memory safe to use here at all: it is not "shared memory", it is one
-- region, named by a capability, given away deliberately.
--
--
-- `periods` is how deep the ring should be, and it is the client's call
-- because the right answer is different for different clients.
--
-- **A shallow ring is low latency and a deep one is safe, and nothing can
-- be both.** A game wants the sound to arrive with the frame that caused
-- it and would rather risk a gap; a video player wants a clock it can hang
-- pictures on and would far rather be 200 ms behind than skip - a gap
-- desynchronises everything after it, where latency is a constant you
-- schedule against. One number cannot serve both, so it stops being a
-- constant.
--
-- Omitted means the default in `audioring.h`, which is what every caller
-- got before this argument existed.
--
function audio.open(name, periods)
  local fmt = audio.format()

  if fmt.period == 0 then return nil, "this machine has no sound device" end

  local cap, at = sys.ring_create(fmt.period, periods)

  if not cap then return nil, tostring(at) end

  -- The capability goes as `send`'s third argument, which is how a window
  -- hands the compositor its surface. It is not a field in the message: a
  -- capability is not data and does not travel as any.
  local r, why = request(OP_OPEN, { name = tostring(name or "sound") }, cap)

  if not r then
    sys.release(cap)

    return nil, tostring(why)
  end

  -- `frames` and `rate` are carried because `delay` is in frames and a
  -- caller wanting milliseconds should not have to ask the machine again
  -- for what this function already had in its hand.
  --
  -- `periods` is the *device's* depth, which is what the server replies
  -- with; `ring_periods` is this ring's, which is a different number and
  -- now a different number per stream.
  local s = setmetatable({ id = r.stream, period = r.period,
                           periods = r.periods, ring = at, cap = cap,
                           frames = fmt.frames, rate = fmt.rate }, stream)

  --
  -- Asked of the ring rather than taken from what was requested.
  --
  -- `ring_create` has limits of its own and quietly uses the default when
  -- it does not like a number, so a client that believed its own argument
  -- would compute a bound that was wrong and never find out. Nothing has
  -- been written yet, so what is free is the whole ring.
  --
  s.ring_periods = s:queued() + s:space()

  return s
end

--------------------------------------------------------------------------
-- Talking to the server, which is C and reads a struct.
--
-- **The layout below is `audioproto.h` written a second time**, and there is
-- no way around it: the server is C and this is Lua, so one holds the shape
-- as a struct and the other as a format string. What can be done is to make
-- a disagreement loud instead of silent - the sizes are asserted here when
-- this library loads, and the server refuses any request that is not
-- exactly the length it expects.
--
-- `serialize.h` argues at length that one implementation on both sides is
-- the point rather than a convenience, and it is right. This is the one
-- place that rule is bent, deliberately, because the alternative is an
-- interpreter inside a server that owes the device a period every 5.8 ms.
-- A check stands where the shared implementation used to.
--------------------------------------------------------------------------

-- struct audio_request: op, stream, gain, balance, muted, master, name[24]
local REQUEST  = "<I4I4i4i4i4i4c24"
local REQ_SIZE = 48

-- struct audio_stream_info: stream, gain, balance, muted, peak, queued, name
local INFO      = "<I4I4i4I4I4I4c24"
local INFO_SIZE = 48

-- struct audio_reply, as far as the list
local REPLY      = "<I4I4I4I4I4I4I4I4I4"
local REPLY_HEAD = 36

assert(#string.pack(REQUEST, 0, 0, 0, 0, 0, 0, "") == REQ_SIZE,
       "audio: the request layout does not match audioproto.h")
assert(#string.pack(INFO, 0, 0, 0, 0, 0, 0, "") == INFO_SIZE,
       "audio: the stream layout does not match audioproto.h")

--
-- An error is a number on the wire and a sentence here.
--
-- A real consequence of a struct protocol, and worth naming rather than
-- hiding: the server detects the fault, this decides what to call it, so
-- the words live with whoever shows them to a person and the server never
-- composes a string.
--
local ERRORS = {
  [1] = "this machine has no sound device",
  [2] = "no such stream",
  [3] = "that is not an audio ring",
  [4] = "too many streams already",
  [5] = "the audio server did not understand that",
}

function request(op, fields, pass)
  local bytes = string.pack(REQUEST, op,
                            fields.stream or 0,
                            fields.gain or -1,
                            fields.balance or -101,
                            fields.muted or -1,
                            fields.master or -1,
                            fields.name or "")

  local reply, why = fs.raw("/dev/audio", bytes, pass)

  if not reply then return nil, tostring(why) end

  if #reply < REPLY_HEAD then
    return nil, "the audio server sent a reply of the wrong size"
  end

  local err, stream, period, periods, master, mixes, starved, late, count =
      string.unpack(REPLY, reply)

  if err ~= 0 then
    return nil, ERRORS[err] or ("audio error " .. tostring(err))
  end

  return { stream = stream, period = period, periods = periods,
           master = master, mixes = mixes, starved = starved,
           late = late, count = count, bytes = reply }
end

--
-- What is playing, for the Mixer and anything else that wants to know.
--
-- `peak` is from the last mix and is measured *before* gain, so a muted
-- stream that is still playing shows a moving meter - which is the question
-- a meter answers: who is sending audio.
--
function audio.streams()
  local r = request(OP_STREAMS, {})

  if not r then return {} end

  local out = {}

  for i = 1, r.count do
    local at = REPLY_HEAD + (i - 1) * INFO_SIZE + 1
    local id, gain, balance, muted, peak, queued, name =
        string.unpack(INFO, r.bytes, at)

    out[i] = { stream = id, gain = gain, balance = balance,
               muted = muted ~= 0, peak = peak, queued = queued,
               name = (name:gsub("%z.*$", "")) }
  end

  return out
end

--
-- What the server says about its own timekeeping.
--
-- `starved` is passes where the device had room and every ring was empty -
-- the clients are behind. `late` is the worst gap between two turns of the
-- server's loop - the server is behind. An underrun is one or the other,
-- and without both there is no telling which.
--
function audio.stats()
  local r = request(OP_STREAMS, {})

  if not r then return nil end

  return { starved = r.starved, late = r.late, mixes = r.mixes }
end

--
-- Change a gain, a balance, a mute, or the master.
--
-- Absent fields are left alone, which is why they are signed on the wire: a
-- `set` that had to send every value would make the Mixer read the server's
-- state back before moving one fader, and two things that must agree is how
-- this system keeps hurting itself.
--
function audio.set(what)
  local r, why = request(OP_SET, {
    stream = what.stream,
    gain = what.gain,
    balance = what.balance,
    muted = (what.muted ~= nil) and (what.muted and 1 or 0) or nil,
    master = what.master,
  })

  if not r then return false, why end

  return true
end

return audio

File: user/servers/audio.c
/* Kosmos. Copyright (c) 2026 Diego Cibils. MIT; see LICENSE. */
/*
 * /dev/audio: the one process that may make a noise, so that several can.
 *
 * **The first server in C**, and the reason is the deadline. The device
 * wants a period every 5.8 milliseconds; a garbage collector in this
 * process would arrive whenever it chose, and `gc_pause_max` is 1.25 ms.
 * A collector and a deadline cannot share a process, and the Lua version of
 * this server met its deadline only because a refactor happened to have
 * removed every allocation from the loop - with nothing to stop the next
 * `print` or table literal putting them back, silently.
 *
 * `SPAWN_AUDIO` grants the device to exactly one process, which is what
 * makes per-application volume possible rather than coincidental: if every
 * program could write to the device, the last writer would win and there
 * would be nothing to turn down.
 *
 * **Samples do not arrive here.** Each client creates a ring in its own
 * memory and hands over a capability at `open`; after that it writes
 * periods and says nothing at all. `audioring.h` has the reasoning, and
 * `CLAUDE.md` has the rule: control by message, data by shared memory.
 */

#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

#include "kosmos.h"
#include "audioring.h"
#include "audioproto.h"
#include "syscall.h"

#define STREAM_MAX  8u          /* one more than AUDIO_LIST_MAX would show */

struct stream {
    bool               open;
    uint32_t           id;
    long               cap;             /* the client's ring, as we name it */
    struct audio_ring *ring;
    int32_t            gain;            /* 0..256, 256 unity */
    int32_t            balance;         /* -100 .. +100 */
    bool               muted;
    uint32_t           peak;            /* from the last mix, before gain */
    char               name[AUDIO_NAME_MAX];
};

static struct stream streams[STREAM_MAX];
static uint32_t      next_id = 1;
static int32_t       master  = 256;

static unsigned      period_bytes;
static unsigned      device_depth;
static unsigned long counter_hz;

/* Diagnostics, reported with `streams`. `starved` is the device having room
 * and every ring being empty - the clients are behind. `late` is the worst
 * gap between two turns of this loop - this server is behind. They are the
 * two halves of a missed deadline and nothing used to measure either. */
static uint32_t      mixes, starved, late;

static struct stream *find(uint32_t id)
{
    unsigned i;

    for (i = 0; i < STREAM_MAX; i++) {
        if (streams[i].open && streams[i].id == id) {
            return &streams[i];
        }
    }

    return NULL;
}

/*
 * One period out of the device, summed from whatever the rings hold.
 *
 * A stream whose ring is empty is skipped rather than waited for: one
 * client falling behind must not take the others down with it. Returns
 * false when nothing was mixed, and the caller must believe it - the Lua
 * version returned "worked" from a count that included empty streams, so it
 * never blocked, span at the top priority band, and took the desktop with
 * it.
 */
static bool refill(void)
{
    static int32_t acc[HAL_SND_PERIOD_BYTES_MAX / 2];
    static int16_t out[HAL_SND_PERIOD_BYTES_MAX / 2];

    unsigned samples = period_bytes / 2;
    unsigned mixed = 0;
    unsigned i, k;

    if (period_bytes == 0) {
        return false;
    }

    if ((unsigned)kosmos_snd_queued() >= device_depth) {
        return false;                   /* the device has all it will hold */
    }

    if (samples > sizeof(acc) / sizeof(acc[0])) {
        samples = (unsigned)(sizeof(acc) / sizeof(acc[0]));
    }

    memset(acc, 0, samples * sizeof(acc[0]));

    for (i = 0; i < STREAM_MAX; i++) {
        struct stream *s = &streams[i];
        const int16_t *in;
        int32_t g, left, right, peak = 0;

        if (!s->open || !audio_ring_valid(s->ring)) {
            continue;
        }

        /* Acquire: having seen this index, the samples behind it are there. */
        if (audio_ring_acquire(&s->ring->write) - s->ring->read == 0) {
            s->peak = 0;
            continue;
        }

        /*
         * Balance is a pan applied to the gain rather than beside it: left
         * of centre attenuates the right channel and vice versa. That is
         * what a balance control does and is not what a *pan* law does - a
         * real pan holds the total power constant and needs a square root.
         * This is the simple one, and saying so beats implying the other.
         */
        g = s->muted ? 0 : ((s->gain * master) / 256);
        left = right = g;

        if (s->balance > 0) {
            left = (g * (100 - s->balance)) / 100;
        } else if (s->balance < 0) {
            right = (g * (100 + s->balance)) / 100;
        }

        in = (const int16_t *)(const void *)audio_ring_slot(s->ring,
                                                            s->ring->read);

        for (k = 0; k < samples; k++) {
            int32_t v = in[k];
            int32_t mag = (v < 0) ? -v : v;

            if (mag > peak) {
                peak = mag;
            }

            /* Even samples are the left channel, odd the right: that is what
             * interleaved stereo means and the only place a balance fits. */
            acc[k] += (v * ((k & 1u) ? right : left)) >> 8;
        }

        s->peak = (uint32_t)peak;
        audio_ring_consumed(s->ring, s->ring->read + 1);
        mixed++;
    }

    if (mixed == 0) {
        return false;
    }

    /*
     * Clipped rather than wrapped. Two streams at full scale sum past what
     * sixteen bits hold, and wrapping turns a peak into the opposite sign -
     * the worst noise a mixer can make.
     */
    for (k = 0; k < samples; k++) {
        int32_t v = acc[k];

        if (v > 32767)  { v = 32767; }
        if (v < -32768) { v = -32768; }

        out[k] = (int16_t)v;
    }

    kosmos_snd_write(out, (unsigned long)samples * 2);
    mixes++;
    return true;
}

static void do_open(const struct audio_request *req, long cap,
                    struct audio_reply *rep)
{
    unsigned i;
    long at;

    if (period_bytes == 0) {
        rep->error = AUDIO_ERR_NO_DEVICE;
        return;
    }

    if (cap < 0) {
        rep->error = AUDIO_ERR_NO_RING;      /* control, but no data path */
        return;
    }

    for (i = 0; i < STREAM_MAX; i++) {
        if (!streams[i].open) {
            break;
        }
    }

    if (i == STREAM_MAX) {
        rep->error = AUDIO_ERR_FULL;
        return;
    }

    at = kosmos_mem_map(cap);

    if (at < 0 || !audio_ring_valid((struct audio_ring *)(uintptr_t)at)) {
        rep->error = AUDIO_ERR_NO_RING;
        return;
    }

    streams[i].open = true;
    streams[i].id = next_id++;
    streams[i].cap = cap;
    streams[i].ring = (struct audio_ring *)(uintptr_t)at;
    streams[i].gain = 256;
    streams[i].balance = 0;
    streams[i].muted = false;
    streams[i].peak = 0;

    /* The name is the client's own and nothing checks it: two copies of one
     * program get two streams called the same thing, which is right - they
     * are two things making a noise and each wants its own fader. */
    memcpy(streams[i].name, req->name, AUDIO_NAME_MAX);
    streams[i].name[AUDIO_NAME_MAX - 1] = '\0';

    rep->error = AUDIO_OK;
    rep->stream = streams[i].id;
    rep->period = period_bytes;
    rep->periods = device_depth;
}

static void do_close(const struct audio_request *req, struct audio_reply *rep)
{
    struct stream *s = find(req->stream);

    if (s == NULL) {
        rep->error = AUDIO_ERR_NO_STREAM;
        return;
    }

    /*
     * The ring was the client's memory and this process only borrowed a view
     * of it. Unmapped before the capability is dropped, because after the
     * drop this process may have no right to name those pages - and dropping
     * first would leave a window where they could be freed underneath a
     * mapping still held here.
     */
    if (s->ring != NULL) {
        unsigned long bytes = AUDIO_RING_DATA
                              + (unsigned long)s->ring->periods
                                * s->ring->period_bytes;

        kosmos_share_unmap((unsigned long)(uintptr_t)s->ring,
                           (bytes + 4095UL) / 4096UL);
    }

    kosmos_cap_drop(s->cap);
    memset(s, 0, sizeof(*s));
    rep->error = AUDIO_OK;
}

static void do_set(const struct audio_request *req, struct audio_reply *rep)
{
    struct stream *s;

    /* Master first, and on its own: a request that only moves the master
     * names no stream and must not be refused for it. */
    if (req->master >= 0) {
        master = (req->master > 256) ? 256 : req->master;
    }

    if (req->stream == 0) {
        rep->error = AUDIO_OK;
        return;
    }

    s = find(req->stream);

    if (s == NULL) {
        rep->error = AUDIO_ERR_NO_STREAM;
        return;
    }

    if (req->gain >= 0)    { s->gain = (req->gain > 256) ? 256 : req->gain; }
    if (req->balance >= -100 && req->balance <= 100) { s->balance = req->balance; }
    if (req->muted >= 0)   { s->muted = (req->muted != 0); }

    rep->error = AUDIO_OK;
}

static void do_streams(struct audio_reply *rep)
{
    unsigned i;

    rep->error = AUDIO_OK;
    rep->master = (uint32_t)master;
    rep->period = period_bytes;
    rep->periods = device_depth;
    rep->mixes = mixes;
    rep->starved = starved;
    rep->late = late;
    rep->count = 0;

    for (i = 0; i < STREAM_MAX && rep->count < AUDIO_LIST_MAX; i++) {
        struct stream *s = &streams[i];
        struct audio_stream_info *o;

        if (!s->open) {
            continue;
        }

        o = &rep->list[rep->count++];
        o->stream = s->id;
        o->gain = (uint32_t)s->gain;
        o->balance = s->balance;
        o->muted = s->muted ? 1u : 0u;
        o->peak = s->peak;
        o->queued = audio_ring_valid(s->ring) ? audio_ring_ready(s->ring) : 0u;
        memcpy(o->name, s->name, AUDIO_NAME_MAX);
    }
}

static void answer(const struct message *in, uint64_t sender, long cap)
{
    struct message out;
    struct audio_reply *rep = (struct audio_reply *)(void *)out.data;
    const struct audio_request *req =
        (const struct audio_request *)(const void *)in->data;

    memset(&out, 0, sizeof(out));
    out.tag = in->tag;
    out.length = (uint32_t)sizeof(*rep);
    out.cap_plus_one = 0;

    if (in->length < sizeof(*req)) {
        rep->error = AUDIO_ERR_BAD_OP;
        (void)kosmos_reply(sender, &out);
        return;
    }

    switch (req->op) {
    case AUDIO_OP_OPEN:    do_open(req, cap, rep); break;
    case AUDIO_OP_CLOSE:   do_close(req, rep);     break;
    case AUDIO_OP_SET:     do_set(req, rep);       break;
    case AUDIO_OP_STREAMS: do_streams(rep);        break;
    default:               rep->error = AUDIO_ERR_BAD_OP; break;
    }

    (void)kosmos_reply(sender, &out);
}

static bool anyone_open(void)
{
    unsigned i;

    for (i = 0; i < STREAM_MAX; i++) {
        if (streams[i].open) {
            return true;
        }
    }

    return false;
}

/*
 * The loop. Never returns.
 *
 * Its shape is the one thing here that took a whole day to get right, so it
 * is worth stating plainly:
 *
 *   - Mix while the device has room and any ring has samples.
 *   - Otherwise wait for a message *or* a tick, whichever comes first.
 *   - Wait for a message alone only when no stream is open at all.
 *
 * The last line is the one that is not obvious. A client with a ring
 * announces nothing - it writes into shared memory and sends no message -
 * so "no ring has samples right now" is true for a moment just after `open`
 * and means nothing. Blocking on that slept through every stream that never
 * spoke. Nothing can begin playing without an `open` first, and that *is* a
 * message, so an idle machine still blocks and an idle desktop is idle.
 */
void audio_server(long endpoint)
{
    struct sysinfo info;
    unsigned long last_turn = 0;
    unsigned long us;

    memset(streams, 0, sizeof(streams));
    memset(&info, 0, sizeof(info));
    (void)kosmos_sysinfo(&info);

    period_bytes = info.audio_period;
    device_depth = info.audio_periods;
    counter_hz = info.counter_hz ? info.counter_hz : 62500000UL;
    us = counter_hz / 1000000UL;

    if (us == 0) {
        us = 1;
    }

    for (;;) {
        struct message msg;
        uint64_t sender = 0;
        long status;
        unsigned long turn = kosmos_ticks();

        /*
         * Only while something is playing. Otherwise the first turn after a
         * stream opens carries the whole idle wait before it - this server
         * blocks until somebody calls `open`, so that gap is however long
         * the machine took to play something, and it read 367 ms for a
         * server that was perfectly healthy.
         */
        if (anyone_open()) {
            if (last_turn != 0) {
                unsigned long gap = (turn - last_turn) / us;

                if (gap > late) {
                    late = (uint32_t)gap;
                }
            }

            last_turn = turn;
        } else {
            last_turn = 0;
        }

        if (refill()) {
            continue;                   /* there may be room for another */
        }

        if ((unsigned)kosmos_snd_queued() < device_depth) {
            starved++;                  /* room, and nothing to put in it */
        }

        /*
         * A deadline on the receive, not a sleep.
         *
         * A server that sleeps on a timer answers nobody while it sleeps, so
         * every caller pays a tick - which cost the Music window 45 ms per
         * pass for twelve round trips that should have been microseconds.
         * This waits for whichever comes first, and the sound interrupt
         * wakes it early through `process_wake_audio`.
         */
        status = kosmos_receive(endpoint, &msg, &sender, 0,
                               (anyone_open() || kosmos_snd_queued() > 0)
                               ? 1UL : 0UL);

        if (status == 0) {
            long cap = (msg.cap_plus_one > 0)
                       ? (long)msg.cap_plus_one - 1 : -1;

            answer(&msg, sender, cap);
        }
    }
}
